File: windows/src/Elevate.Core/Providers/AzureApprovalProvider.cs
using System.Text.Json;
using Elevate.Core.Auth;
using Elevate.Core.Models;
using Elevate.Core.Networking;
using Elevate.Core.Support;

namespace Elevate.Core.Providers;

/// <summary>Azure resource role activation requests awaiting the signed-in user's approval, through ARM.</summary>
public sealed class AzureApprovalProvider : IApprovalProvider
{
    internal const string ListApiVersion = "2020-10-01";
    internal const string ApprovalApiVersion = "2021-01-01-preview";

    private readonly GraphTransport _transport;

    public AzureApprovalProvider(IHttpClient http, ITokenProvider tokens)
        => _transport = new GraphTransport(http, tokens, GraphTransport.MapArmError);

    public RoleScopeKind Kind => RoleScopeKind.AzureResource;

    public IReadOnlyList<string> Scopes => Auth.Scopes.ArmAll;

    // MARK: Wire models

    private sealed record RequestProperties(
        string? RoleDefinitionId,
        string? PrincipalId,
        string? RequestType,
        string? Status,
        string? ApprovalId,
        string? Justification,
        DateTimeOffset? CreatedOn,
        ScheduleInfo? ScheduleInfo,
        AzureResourceProvider.Expanded? ExpandedProperties);

    private sealed record Request(string Name, RequestProperties Properties);

    private sealed record StageProperties(string? ReviewResult, string? Status);

    private sealed record Stage(string? Name, string? Id, StageProperties? Properties);

    private sealed record ApprovalProperties(IReadOnlyList<Stage>? Stages);

    private sealed record Approval(ApprovalProperties? Properties);

    // MARK: Reads

    public async Task<IReadOnlyList<ApprovalRequest>> PendingApprovalsAsync(
        Identity identity, TenantContext tenant, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        ArgumentNullException.ThrowIfNull(tenant);

        var url = AzureResourceProvider.ArmUrl(
            "providers/Microsoft.Authorization/roleAssignmentScheduleRequests",
            ListApiVersion, new Dictionary<string, string> { ["$filter"] = "asApprover()" });
        var items = await AzureResourceProvider.ListAllAsync<Request>(_transport, identity, tenant.TenantId, url, Scopes, ct)
            .ConfigureAwait(false);

        return [.. items
            .Where(r => r.Properties.Status == "PendingApproval")
            .Select(r =>
            {
                var expanded = r.Properties.ExpandedProperties;
                return new ApprovalRequest(
                    r.Name, tenant.Key, Kind, GraphApprovals.Action(r.Properties.RequestType),
                    targetName: expanded?.RoleDefinition?.DisplayName ?? r.Properties.RoleDefinitionId ?? r.Name,
                    scopeCaption: AzureResourceProvider.Caption(expanded),
                    requesterName: expanded?.Principal?.DisplayName ?? r.Properties.PrincipalId ?? "Unknown",
                    justification: r.Properties.Justification,
                    requestedDuration: GraphApprovals.Duration(r.Properties.ScheduleInfo),
                    createdAt: r.Properties.CreatedOn,
                    decisionRef: r.Properties.ApprovalId);
            })];
    }

    // MARK: Decision

    /// <summary>The stage to decide: the one still awaiting review, else the first.</summary>
    private static string PickStage(IReadOnlyList<Stage> stages)
    {
        var pick = stages.FirstOrDefault(s => string.Equals(s.Properties?.ReviewResult, "NotReviewed", StringComparison.OrdinalIgnoreCase))
            ?? stages.FirstOrDefault();
        return pick?.Name ?? pick?.Id?.Split('/').LastOrDefault() ?? throw GraphApprovals.NoStep();
    }

    public async Task DecideAsync(ApprovalRequest request, bool approve, string justification, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        var approvalId = request.DecisionRef ?? throw GraphApprovals.NoStep();
        var tenantId = request.TenantKey.TenantId;
        var basePath = $"providers/Microsoft.Authorization/roleAssignmentApprovals/{approvalId}";
        var response = await _transport.GetAsync(
            identity, tenantId, AzureResourceProvider.ArmUrl(basePath, ApprovalApiVersion), Scopes, ct).ConfigureAwait(false);
        var approval = JsonSerializer.Deserialize<Approval>(response.Body, GraphJson.Options);
        var stage = PickStage(approval?.Properties?.Stages ?? []);
        var url = AzureResourceProvider.ArmUrl($"{basePath}/stages/{stage}", ApprovalApiVersion);
        var decision = GraphApprovals.Decision(approve, justification);
        try
        {
            await _transport.PatchAsync(identity, tenantId, url, Scopes, JsonSerializer.SerializeToUtf8Bytes(decision), ct)
                .ConfigureAwait(false);
        }
        catch (PimException e) when (e.Kind == PimErrorKind.Unexpected && e.Status == 400)
        {
            // Some ARM versions want the decision wrapped in `properties`; the flat form the docs show comes first.
            var wrapped = new Dictionary<string, Dictionary<string, string>> { ["properties"] = decision };
            await _transport.PatchAsync(identity, tenantId, url, Scopes, JsonSerializer.SerializeToUtf8Bytes(wrapped), ct)
                .ConfigureAwait(false);
        }
    }
}

File: macos/Sources/ElevateCore/Providers/AzureApprovalProvider.swift
import Foundation

/// Azure resource role activation requests awaiting the signed-in user's approval, through ARM.
public struct AzureApprovalProvider: ApprovalProvider {
    public let kind: RoleScopeKind = .azureResource
    public let scopes = ArmScopes.all
    let transport: GraphTransport

    public init(http: any HTTPClient, tokens: any TokenProviding) {
        transport = GraphTransport(http: http, tokens: tokens, mapper: GraphTransport.mapArmError)
    }

    // MARK: Wire models

    struct RequestProperties: Decodable {
        let roleDefinitionId: String?
        let principalId: String?
        let requestType: String?
        let status: String?
        let approvalId: String?
        let justification: String?
        let createdOn: Date?
        let scheduleInfo: ScheduleInfo?
        let expandedProperties: AzureResourceProvider.Expanded?
    }
    struct Request: Decodable { let name: String; let properties: RequestProperties }
    struct Page<T: Decodable>: Decodable { let value: [T]; let nextLink: String? }
    struct StageProperties: Decodable { let reviewResult: String?; let status: String? }
    struct Stage: Decodable { let name: String?; let id: String?; let properties: StageProperties? }
    struct ApprovalProperties: Decodable { let stages: [Stage]? }
    struct Approval: Decodable { let properties: ApprovalProperties? }

    static let listAPIVersion = "2020-10-01"
    static let approvalAPIVersion = "2021-01-01-preview"

    // MARK: Reads

    public func pendingApprovals(identity: Identity, tenant: TenantContext) async throws -> [ApprovalRequest] {
        var next: URL? = try AzureResourceProvider.armURL("providers/Microsoft.Authorization/roleAssignmentScheduleRequests",
                                         apiVersion: Self.listAPIVersion, query: ["$filter": "asApprover()"])
        var items: [Request] = []
        while let current = next {
            let r = try await transport.get(identity: identity, tenantId: tenant.tenantId, url: current, scopes: scopes)
            let page = try GraphJSON.decoder.decode(Page<Request>.self, from: r.body)
            items += page.value
            next = page.nextLink.flatMap(URL.init(string:))
        }
        return items.compactMap { r -> ApprovalRequest? in
            guard r.properties.status == "PendingApproval" else { return nil }
            let expanded = r.properties.expandedProperties
            return ApprovalRequest(
                id: r.name, tenantKey: tenant.id, kind: kind, action: GraphApprovals.action(r.properties.requestType),
                targetName: expanded?.roleDefinition?.displayName ?? r.properties.roleDefinitionId ?? r.name,
                scopeCaption: AzureResourceProvider.caption(expanded),
                requesterName: expanded?.principal?.displayName ?? r.properties.principalId ?? "Unknown",
                justification: r.properties.justification,
                requestedDuration: r.properties.scheduleInfo?.expiration?.duration.flatMap(ISO8601Duration.parse),
                createdAt: r.properties.createdOn, decisionRef: r.properties.approvalId)
        }
    }

    // MARK: Decision

    /// The stage to decide: the one still awaiting review, else the first.
    static func stage(_ stages: [Stage]) throws -> String {
        let pick = stages.first { $0.properties?.reviewResult?.caseInsensitiveCompare("NotReviewed") == .orderedSame } ?? stages.first
        guard let name = pick?.name ?? pick?.id?.components(separatedBy: "/").last else {
            throw PIMError.unexpected(status: 0, body: "No approval step to decide")
        }
        return name
    }

    public func decide(_ request: ApprovalRequest, approve: Bool, justification: String, identity: Identity) async throws {
        guard let approvalId = request.decisionRef else {
            throw PIMError.unexpected(status: 0, body: "No approval step to decide")
        }
        let tenantId = request.tenantKey.tenantId
        let base = "providers/Microsoft.Authorization/roleAssignmentApprovals/\(approvalId)"
        let r = try await transport.get(identity: identity, tenantId: tenantId,
                                        url: try AzureResourceProvider.armURL(base, apiVersion: Self.approvalAPIVersion), scopes: scopes)
        let approval = try GraphJSON.decoder.decode(Approval.self, from: r.body)
        let stage = try Self.stage(approval.properties?.stages ?? [])
        let url = try AzureResourceProvider.armURL("\(base)/stages/\(stage)", apiVersion: Self.approvalAPIVersion)
        let decision = ["reviewResult": approve ? "Approve" : "Deny", "justification": justification]
        do {
            _ = try await transport.patch(identity: identity, tenantId: tenantId, url: url, scopes: scopes,
                                          body: try JSONSerialization.data(withJSONObject: decision))
        } catch PIMError.unexpected(status: 400, _) {
            // Some ARM versions want the decision wrapped in `properties`; the flat form the docs show comes first.
            _ = try await transport.patch(identity: identity, tenantId: tenantId, url: url, scopes: scopes,
                                          body: try JSONSerialization.data(withJSONObject: ["properties": decision]))
        }
    }
}
